### PR TITLE
ISPN-12518 Possible deadlock if an entry is replaced at the same time…

### PR DESCRIPTION
**Pull request by wburns**
… expiration take effect

* Ensure backup writes don't check expiration

https://issues.redhat.com/browse/ISPN-12518

### CODE REVIEW

File: core/src/main/java/org/infinispan/commands/triangle/BackupWriteCommand.java [hunk 1]
```diff
@@ -46,7 +46,8 @@
    public final CompletionStage<?> invokeAsync(ComponentRegistry componentRegistry) {
       WriteCommand command = createWriteCommand();
       command.init(componentRegistry);
-      command.setFlagsBitSet(flags);
+      // Mark the command as a backup write
+      command.setFlagsBitSet(flags | FlagBitSets.BACKUP_WRITE);
       command.addFlags(FlagBitSets.SKIP_LOCKING);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/commands/triangle/BackupWriteCommand.java:51]: nit: can you merge `FlagBitSets.SKIP_LOCKING` to the line above and remove this line?

File: core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java [hunk 2]
```diff
@@ -432,4 +439,32 @@ private void testMaxIdleNodeDies(Function<Cache<Object, String>, Object> keyToUs
 
       assertNotNull(cache1.get(key));
    }
+
+   public static final int LIFESPAN = 64;
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:443]: unused. can be removed.

File: core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java [hunk 3]
```diff
@@ -432,4 +439,32 @@ private void testMaxIdleNodeDies(Function<Cache<Object, String>, Object> keyToUs
...
+
+   public void testPrimaryNotExpiredButBackupWas() throws InterruptedException, ExecutionException, TimeoutException {
+      if (transactional || cacheMode != CacheMode.DIST_SYNC) {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:446]: so, isn't the replicated cache affected by this?

File: core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java [hunk 4]
```diff
@@ -432,4 +439,32 @@ private void testMaxIdleNodeDies(Function<Cache<Object, String>, Object> keyToUs
...
+
+         incrementAllTimeServices(11, TimeUnit.SECONDS);
+
+         blockedRequest.send();
+
+         assertEquals(value, result.get(10, TimeUnit.SECONDS));
+      } finally {
+         controlledRpcManager.revertRpcManager();
+      }
+   }
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:469]: for completeness :) 
```suggestion
   assertEquals(value + "-expire-backup", cache0.get(key));
   assertEquals(value + "-expire-backup", cache1.get(key));
   assertEquals(value + "-expire-backup", cache2.get(key));
   }
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/commands/triangle/BackupWriteCommand.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/commands/triangle/BackupWriteCommand.java:51]: Ah yes, didn't notice 👍

File: core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java (hunk 4 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:469]: 👍

File: core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java (hunk 3 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:446]: It is, but the test wasn't utilizing it :(

Comment (wburns) [reply core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:446]: I will try to tweak it to add it, requires slightly different rpc handling.

Comment (wburns) [reply core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java:446]: REPL is broken atm, have to figure out how to fix it still.

**Comment by wburns** [comment]
Updated.

File: core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java [hunk 5]
```diff
@@ -270,6 +270,8 @@ protected Object primaryReturnHandler(InvocationContext ctx, AbstractDataWriteCo
       }
       MapResponseCollector collector = MapResponseCollector.ignoreLeavers(isReplicated, owners.size());
       RpcOptions rpcOptions = rpcManager.getSyncRpcOptions();
+      // Mark the command as a backup write so it can skip some checks
+      command.addFlags(FlagBitSets.BACKUP_WRITE);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java:274]: I think you have to remove the flag after sending (after line 280 below). if the command retries (topology changed), the command will have this flag on the primary owner.

Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java:274]: Makes sense, let me take a closer look.

**Comment by wburns** [comment]
Updated, thanks again @pruivo !

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
@wburns possible related failure in CI?
```
org.infinispan.expiration.impl.ClusterExpirationFunctionalTest.testMaxIdleExpiredOnBackupOwner[REPL_SYNC, tx=true, locking=PESSIMISTIC, HEAP]
```

**Comment by wburns** [comment]
> @wburns possible related failure in CI?
> 
> ```
> org.infinispan.expiration.impl.ClusterExpirationFunctionalTest.testMaxIdleExpiredOnBackupOwner[REPL_SYNC, tx=true, locking=PESSIMISTIC, HEAP]
> ```

Checking.

**Comment by wburns** [comment]
@pruivo It doesn't appear so as the test failed in the same way for scattered cache on master. https://ci.infinispan.org/job/Infinispan/job/master/lastCompletedBuild/testReport/org.infinispan.expiration.impl/ClusterExpirationFunctionalTest/testMaxIdleExpiredOnBackupOwner_SCATTERED_SYNC__tx_false__OBJECT_/

**Comment by pruivo** [comment]
integrated! thanks @wburns !

**Pull request merged**